test/rules/no_self_assignments_test.dart

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(NoSelfAssignmentsTest);
+  });
+}
+
+@reflectiveTest
+class NoSelfAssignmentsTest extends LintRuleTest {
+  @override
+  String get lintRule => 'no_self_assignments';
+
+  test_fieldAssignment() async {
+    await assertDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int x) {
+    x = x;
+  }
+}
+''', [lint(41, 5)]);
+  }
+
+  test_fieldAssignmentDifferentVar() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int y) {
+    x = y;
+  }
+}
+''');
+  }
+
+  test_fieldAssignmentExplicit() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int x) {
+    this.x = x;
+  }
+}
+''');
+  }
+
+  test_fieldAssignmentExplicitSameVar() async {
+    await assertDiagnostics(r'''
+class C {
+  int x = 5;
+
+  void update(C other) {
+    other.x = other.x;
+  }
+}
+''', [lint(53, 17)]);
+  }
+
+  test_fieldAssignmentThisAndDifferentTarget() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x = 5;
+
+  void update(C other) {
+    this.x = other.x;
+  }
+}
+''');
+  }
+
+  test_fieldAssignmentDifferentTargets() async {
+    await assertNoDiagnostics(r'''
+class C {
+  String hello = 'ok';
+}
+
+void test(C one, C two) {
+  one.hello = two.hello;
+}
+''');
+  }
+
+  test_fieldInitialization() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x;
+
+  C(int x) : x = x;
+}
+''');
+  }
+
+  test_propertyAssignment() async {
+    await assertDiagnostics(r'''
+class C {
+  int _x = 5;
+
+  int get x => _x;
+
+  set x(int x) {
```

Comment:
I feel like this came up in discussion but I can't recall and maybe this is why you call out "direct" effects above but I'm guessing we don't mind reporting in a case where a setter is side-effecting?  That's an anti-pattern too but slightly different than variable self-assignment.

Reply:
I don't think I've seen a discussion on this myself. It's just coming from simple considerations:

* I didn't distinguish between normal fields and setters in the initial implementation. And after seeing the results, I thought that they were both worth flagging and refactoring.
* The "caller site" looks exactly the same (just `x = x`) and lint rules should be simple to understand and resistant to code edits in distant locations.